SPECS/tdnf/virtual-repo-snapshot.patch

```diff
@@ -0,0 +1,906 @@
+@@ -9,6 +9,7 @@
+ #pragma once
+ 
+ #include <curl/curl.h>
++#include <time.h>
+ 
+ #ifdef __cplusplus
+ extern "C" {
+@@ -257,6 +258,7 @@ typedef struct _TDNF_CONF
+     int nCheckUpdateCompat;
+     int nDistroSyncReinstallChanged;
+     char* pszRepoDir;
++    char* pszRepoEpochTimeLimit;
+     char* pszCacheDir;
+     char* pszPersistDir;
+     char* pszProxy;
+diff --git a/solv/defines.h b/solv/defines.h
+index 38f5ab1..ebc7bd0 100644
+--- a/solv/defines.h
++++ b/solv/defines.h
+@@ -18,4 +18,32 @@
+         }                                                          \
+     } while(0)
+ 
++typedef struct {
++    // frequently changed values
++    char * pszElementBuffer;
++    int nBufferLen;
++    int nInPackage;
++    int nPrintPackage;
++    int nTimeFound;
++
++    // managed values
++    int nBufferMaxLen;
++    int nDepth;
++    int nPrevElement; // enum 0 -> start, 1 -> data, 2 -> end
++
++    //set and forget on creation
++    time_t nSearchTime;
++    FILE * pbOutfile;
++} TDNFFilterData;
++
++#define TDNF_MAX_FILTER_INPUT_THRESHOLD 10000000
++
++#define BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError) \
++    do {                                                           \
++        if (dwError)                                               \
++        {                                                          \
++            goto error;                                            \
++        }                                                          \
++    } while(0)
++
+ #endif /* __SOLV_DEFINES_H__ */
+diff --git a/solv/includes.h b/solv/includes.h
+index 2ab0c5c..1e39ecb 100644
+--- a/solv/includes.h
++++ b/solv/includes.h
+@@ -10,7 +10,6 @@
+ #include <unistd.h>
+ #include <stdbool.h>
+ #include <sys/types.h>
+-#include <sys/stat.h>
+ #include <dirent.h>
+ 
+ // libsolv
+@@ -44,4 +43,9 @@
+ #include "../history/history.h"
+ #include "prototypes.h"
+ 
++#include <expat.h>
++#include <stdlib.h>
++#include <time.h>
++#include <string.h>
++
+ #endif /* __SOLV_INCLUDES_H__ */
+diff --git a/solv/prototypes.h b/solv/prototypes.h
+index 2633b5e..0f5df8d 100644
+--- a/solv/prototypes.h
++++ b/solv/prototypes.h
+@@ -517,7 +517,8 @@ SolvReadYumRepo(
+     const char *pszPrimary,
+     const char *pszFilelists,
+     const char *pszUpdateinfo,
+-    const char *pszOther
++    const char *pszOther,
++    char * pszRepoEpochLimit
+     );
+ 
+ uint32_t
+diff --git a/solv/tdnfrepo.c b/solv/tdnfrepo.c
+index c27b907..aef6c61 100644
+--- a/solv/tdnfrepo.c
++++ b/solv/tdnfrepo.c
+@@ -9,6 +9,522 @@
+ 
+ #include "includes.h"
+ 
++// #### XML FILTER CODE ####
++void checkAndResizeBuffer(char ** pszCharBuffer, int * nBufferMaxLen, int nLengthToAdd) {
++    uint32_t dwError = 0;
++    int nBufferContentLen = strlen(*pszCharBuffer);
++    while (nBufferContentLen + nLengthToAdd >= *nBufferMaxLen)
++    {
++        *nBufferMaxLen *= 2;
++        char * pszTempCharBuffer = NULL;
++        pszTempCharBuffer = realloc(*pszCharBuffer, *nBufferMaxLen);
```

Comment:
This realloc is called in a loop.  I don't think you want to do that.

Reply:
Yes this realloc is called within a loop. This is a way minimize the number of calls to realloc in scenarios where this function will be called back to back (character linting or adding large character chunks to the cached package element). This allows the buffer to be overestimated, rather than being realloced exactly each time potentially increasing the number of required calls to the function.

Reply:
If the realloc fails, the function will simply bail as if any other realloc would bail, causing the program to fail.

Reply:
per some more discussion on a call, the realloc has been moved outside of the loop and only the size recalculation is done inside of the loop, to reduce the number of calls to realloc even further.